crates/core/tedge_agent/src/software_manager/actor.rs

```diff
@@ -229,6 +243,48 @@ impl SoftwareManagerActor {
         self.output_sender.send(response.into()).await?;
 
         self.state_repository.clear().await?;
+        Self::detect_self_update()?;
+        Ok(())
+    }
+
+    fn detect_self_update() -> Result<(), SoftwareManagerError> {
+        info!("Checking if tedge got self updated");
+        let current_running_version = env!("CARGO_PKG_VERSION");
+        info!("Current running version: {}", current_running_version);
+        if let Ok(executable_path) = std::env::current_exe() {
+            let agent_binary_path = executable_path.parent().unwrap().join(TEDGE_AGENT);
+            if let Ok(output) = Command::new(agent_binary_path).args(["--version"]).output() {
+                if output.status.success() {
+                    let version_output = String::from_utf8_lossy(&output.stdout);
+                    let version_output_split: Vec<&str> =
+                        version_output.split_whitespace().collect();
+                    if let ["tedge-agent", installed_binary_version] =
+                        version_output_split.as_slice()
+                    {
+                        info!("Installed binary version: {}", installed_binary_version);
+                        if current_running_version != *installed_binary_version {
+                            info!("Self update detected. Requesting shutdown...");
+                            return Err(SoftwareManagerError::Shutdown);
+                        }
```

Comment:
I would return a `bool`, just telling that a self update has been detected. The fact that a restart is needed is a solution not a diagnostic.